pkg/reconciler/kubernetes/tektonhub/tektonhub.go

```diff
@@ -0,0 +1,595 @@
+}
+
+var (
+	apiConfigMapName string = "api"
+	errKeyMissing    error  = fmt.Errorf("secret doesn't contains all the keys")
+	namespace        string = ""
+	// Check that our Reconciler implements controller.Reconciler
+	_ tektonhubconciler.Interface = (*Reconciler)(nil)
+	_ tektonhubconciler.Finalizer = (*Reconciler)(nil)
+)
+
+const (
+	dbInstallerSet          = "DbInstallerSet"
+	dbMigrationInstallerSet = "DbMigrationInstallerSet"
+	apiInstallerSet         = "ApiInstallerSet"
+
+	releaseVersionKey  = "operator.tekton.dev/release-version"
+	createdByKey       = "operator.tekton.dev/created-by"
+	targetNamespaceKey = "operator.tekton.dev/target-namespace"
+	createdByValue     = "TektonHub"
+)
+
+// FinalizeKind removes all resources after deletion of a TektonHub.
+func (r *Reconciler) FinalizeKind(ctx context.Context, original *v1alpha1.TektonHub) pkgreconciler.Event {
+	logger := logging.FromContext(ctx)
+
+	installerSets := original.Status.HubInstallerSet
+	if len(installerSets) == 0 {
+		return nil
+	}
+
+	for _, value := range installerSets {
+		err := r.operatorClientSet.OperatorV1alpha1().TektonInstallerSets().Delete(ctx, value, metav1.DeleteOptions{})
+		if err != nil && !apierrors.IsNotFound(err) {
+			return err
+		}
+	}
+
+	// List all TektonHub to determine if cluster-scoped resources should be deleted.
+	ths, err := r.operatorClientSet.OperatorV1alpha1().TektonHubs().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to list all TektonHubs: %w", err)
+	}
+
+	for _, th := range ths.Items {
+		if th.GetDeletionTimestamp().IsZero() {
+			// Not deleting all TektonHubs. Nothing to do here.
+			return nil
+		}
+	}
+
+	if err := r.extension.Finalize(ctx, original); err != nil {
+		logger.Error("Failed to finalize platform resources", err)
+	}
+	logger.Info("Deleting cluster-scoped resources")
+	manifest, err := r.installed(ctx, original)
+	if err != nil {
+		logger.Error("Unable to fetch installed manifest; no cluster-scoped resources will be finalized", err)
+		return nil
+	}
+	if err := common.Uninstall(ctx, manifest, nil); err != nil {
+		logger.Error("Failed to finalize platform resources", err)
+	}
+	return nil
+}
+
+// ReconcileKind compares the actual state with the desired, and attempts to
+// converge the two.
+func (r *Reconciler) ReconcileKind(ctx context.Context, th *v1alpha1.TektonHub) pkgreconciler.Event {
+	logger := logging.FromContext(ctx)
+	th.Status.InitializeConditions()
+	th.Status.ObservedGeneration = th.Generation
+	namespace = th.Spec.GetTargetNamespace()
+
+	if namespace == "" {
+		namespace = os.Getenv("DEFAULT_TARGET_NAMESPACE")
+		th.Spec.TargetNamespace = namespace
+	}
+
+	_, err := r.kubeClientSet.CoreV1().Namespaces().Get(ctx, namespace, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			if err := common.CreateTargetNamespace(map[string]string{}, th, r.kubeClientSet); err != nil {
+				return err
+			}
+		}
+		return err
+	}
```

Comment:
extract this into a function, these checks are a bit low level from the reconciler loop context.